pip/qsharp/_qsharp.py

```diff
@@ -215,10 +218,20 @@ def on_save_events(output: Output) -> None:
         # Append the output to the last shot's output list
         results[-1]["events"].append(output)
 
-    for _ in range(shots):
+    # Run the first shot using `run`, which will compile the entry expression
+    results.append({"result": None, "events": []})
+    run_results = get_interpreter().run(
+        entry_expr, on_save_events if save_events else print_output
+    )
+    results[-1]["result"] = run_results
+    if on_result:
+        on_result(results[-1])
+
+    # Run the remaining shots using `rerun`, which will reuse the compiled entry expression
+    for _ in range(shots - 1):
```

Comment:
Rather than duplicate all the logic and vars, why not have it all in the for loop, and just call `run` if iteration 1, else `rerun`?

Reply:
Using your suggestion on making the expression optional in `run` should simplify this.